feat: extend MRSV to 1.63

aljazerzen opened the pull request:
Closes #https://github.com/aj-bagwell/clio/issues/13

Adds alternative implementation of `add_extension` function, which is slightly less efficient as it clones the path instead of mutating it.

This does add a dependency to `rustversion` for conditional compilation depending on Rust version. If that is not desired, `add_extension` function can be replaced completely.

aljazerzen committed:
feat: extend MRSV to 1.63

aj-bagwell commented on src/path.rs:
```diff
@@ -230,6 +231,32 @@ impl ClioPath {
+    #[rustversion::before(1.70)]
+    pub fn add_extension<S: AsRef<OsStr>>(&mut self, extension: S) -> bool {
+        if self.file_name().is_some() && !self.ends_with_slash() {
+            let mut existing: OsString = self.extension().unwrap_or_default().to_os_string();
```
If there is no extension then it will default to `""` which will add an extention of `".ext"` which will cause it to add two dots.

Could you add a doc test along the lines of
```
let mut p = ClioPath::new("/tmp/log")?;
p.add_extension("gz");
assert_eq!(ClioPath::new("/tmp/log.gz")?, p);
```

aj-bagwell commented on src/path.rs:
```diff
@@ -230,6 +231,32 @@ impl ClioPath {
+            existing.push(".");
+            existing.push(extension);
+            self.with_path_mut(|path| path.set_extension(existing));
+            true
```
This also made me realize there is a bug in the current implementation where for stdin it will not update the path but erroneously return `true` 

That could be fixed by removing all the `true` and returning the result of `with_path_mut`

aj-bagwell commented on src/path.rs:
````diff
@@ -217,6 +217,7 @@ impl ClioPath {
     /// assert_eq!(ClioPath::new("/tmp/log.txt.gz")?, p);
     /// # Ok::<(), clio::Error>(())
     /// ```
+    #[rustversion::since(1.70)]
````
I'm not sure it is worth adding the dependency and second implementation just to avoid one allocation, but I am open to being persuaded otherwise.

aj-bagwell commented on Cargo.toml:
```diff
@@ -9,6 +9,7 @@ repository = "https://github.com/aj-bagwell/clio"
 documentation = "https://docs.rs/clio"
 readme = "README.md"
 edition = "2021"
+rustversion = "1.63.0"
```
Is this supposed to be [`rust-version`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field)?

Also I know I am probably getting gready but it would be nice to add a MRSV check to the CI.

The config can probably steal the config from [prql](https://github.com/PRQL/prql/blob/26f6d5f4679beb43b4f9e7e0813d3f65f29e6a72/.github/workflows/test-all.yaml#L130)

epage commented on Cargo.toml:
```diff
@@ -30,6 +31,7 @@ cfg-if = "1.0.0"
 tempfile = "3.3.0"
 walkdir = "2.3.3"
 atty = "0.2.14"
+rustversion = "1.0"
```
This is a very expensive dependency
- its a proc macro
- it has a build script that has to call out to `rustc` (granted, thats better than feature detection which requires compiling samples of Rust code)

epage commented on src/path.rs:
````diff
@@ -230,6 +231,32 @@ impl ClioPath {
         }
     }
 
+    /// Adds an extension to the end of the [`self.file_name`](Path::file_name).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use clio::ClioPath;
+    ///
+    /// let mut p = ClioPath::new("/tmp/log.txt")?;
+    ///
+    /// p.add_extension("gz");
+    /// assert_eq!(ClioPath::new("/tmp/log.txt.gz")?, p);
+    /// # Ok::<(), clio::Error>(())
+    /// ```
+    #[rustversion::before(1.70)]
+    pub fn add_extension<S: AsRef<OsStr>>(&mut self, extension: S) -> bool {
````
If this moves forward, I would recommend having a single public function with a single piece of documentation that then calls into a shared function.  That shred function shouldn't accept any generic parameters (in general, any function that accepts generic parameters like `AsRef` should do this)

aljazerzen committed:
fix

aljazerzen committed:
add a check to gh actions

aljazerzen commented:
I've removed the `rustversion` dep and the old function.

I've also updated it to pass all tests and return correct result. There a bit of duplication, but it was required due to different types.

aj-bagwell commented on .github/workflows/rust.yml:
```diff
@@ -44,3 +44,9 @@ jobs:
+
+    - uses: baptiste0928/cargo-install@v2
+      with:
+        crate: cargo-msrv
```
`cargo-msrv` is failing to install on windows (looks like [an issue with the version of tokio](https://stackoverflow.com/questions/76173195/my-rust-project-is-broken-all-of-the-sudden)). If you know how to get cargo to install the cargo-msrv with the right tokio & ntapi versions that would be ideal, if not just move this to the `doc` job so it only runs on ubuntu.

epage replied:
Personally, I verify MSRV by running regular cargo commands.  I then have [renovatebot](https://github.com/clap-rs/clap/blob/master/.github/renovate.json5#L8) update MSRV for me

aljazerzen replied:
Yeah, cargo-msrv is not really required here.

Feel free to remove.

aj-bagwell commented on src/path.rs:
```diff
@@ -204,7 +204,7 @@ impl ClioPath {
     }
 
-    /// Adds an extention to the end of the [`self.file_name`](Path::file_name).
+    /// Adds an extension to the end of the [`self.file_name`](Path::file_name).
```
For once it looks like I spelled something correctly, it really is `extension` e.g. [`PathBuf::set_extension`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.set_extension).

aljazerzen replied:
Did you highlight the correct line? I've changed the spelling to "extension".

aj-bagwell replied:
you are right I should go get my eyes tested to see if I am colour blind.

aj-bagwell replied:
I don't need my eyes tested, I was compairing against master where you have already fixed it!

aljazerzen committed:
run msrv check on docs task only

aljazerzen committed:
fix

aj-bagwell commented:
Thanks for all your work, I have merged it and published it in version 0.3.3
